source/common/stats/recent_lookups.cc

```diff
@@ -0,0 +1,40 @@
+#include "common/stats/recent_lookups.h"
+
+#include <functional>
+#include <utility>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/str_join.h"
+
+namespace Envoy {
+namespace Stats {
+
+namespace {
+constexpr size_t Capacity = 10;
+} // namespace
+
+void RecentLookups::lookup(absl::string_view str) {
+  ++total_;
+  if (queue_.size() >= Capacity) {
+    queue_.pop_back();
+  }
+  queue_.push_front(std::string(str));
+}
+
+/**
+ * Calls fn(item, timestamp) for each of the remembered lookups.
+ *
+ * @param fn The function to call for every recently looked up item.
+ */
+void RecentLookups::forEach(IterFn fn) const {
+  absl::flat_hash_map<absl::string_view, uint64_t> counts;
+  for (const std::string& item : queue_) {
+    ++counts[item];
```

Comment:
I'm a little confused here. Isn't the idea to keep counts of things that we keep looking up? With a capacity of 10 are you hoping to just get lucky if the same thing keeps getting looked up over and over again?

Reply:
Sorry stepping back I'm still confused though. AFAICT the implementation of lookup always just pops the back and pushes to the front.

At the time of iterating over the dequeue I see that we are counting the items, but wouldn't it be better to keep a running count of the lookups as I'm not sure that counting at the time of admin fetch is all that useful as that could be done just as well by a script? I realize we are going in circles a bit, but couldn't this just be a parallel LRU and map, where the map is from string -> count/LRU entry, and then if there is no map entry we just pop off the back? If we don't want to do this, can you flesh out the comments in recent_lookups.h with well specified properties of what we are storing, runtime complexity, tradeoffs, etc.? I think that would help the discussion.

Reply:
I have a branch implementing the LRU: https://github.com/jmarantz/envoy/blob/8cdf5fc83a2fa43213f6e555452c49808e4fe4bc/source/common/stats/recent_lookups.h#L67 and https://github.com/jmarantz/envoy/blob/8cdf5fc83a2fa43213f6e555452c49808e4fe4bc/source/common/stats/recent_lookups.cc#L25

It's just >20x slower than the deque (measured with a microbenchmark) and the main difference at the end of the day is that for a given memory commitment you will see a lot more of the violating stats with the LRU.

The cost is that anyone running an experiment with the real symbol tables will be holding the global mutex for much longer due to the LRU, and thus increasing contention on real traffic.

If the goal is to just find and eliminate all the contending stats then I think the LRU might make the iterations faster by capturing more violations. But I'I think it's not worth it as it makes the experiments more expensive, and we should be able to converge based on data from the deque.

Take for example the http2 stats I found we were looking up by name on every request in https://github.com/envoyproxy/envoy/issues/8324 -- the 'total' counter told us there are 12, but I only list 10 of them because that's the size of the deque. I'm guessing there's one code-change to fix all 12, but if there wasn't we could just do another run to find out the 2 that fell off. In this particular case the LRU wouldn't have mattered because each was looked up just one time per request.

The LRU probably will help narrow things down faster in some cases.

Anyway that's my reasoning; I'm OK if you prefer I go with LRU anyway in which case I'll document the perf concern, or I could add this rationale in a comment and stay with this one.

Reply:
What if the lookup accounting is off by default and possibly enabled via admin endpoint and/or CLI flag?

We can have a stat for total lookups, and then if the operator actually wants to debug in depth they could enable the lookup debugging? WDYT?

Reply:
SGTM will implement later today during code therapy.

Reply:
qq; is the 'stat' as I have it; just a line printed from `/stats?recentlookups` sufficient? Or do we want it to be intgrated into `/stats`?

Reply:
IMO it should be integrated into `/stats` as an actual `server` stat, as this is what I think people would want to monitor?

Reply:
so probably I want to make a real stat for this, which I can probably inject via lambda or something.

Reply:
I have made the LRU implementation but I didn't make the total a real stat yet.

I wanted to get a taste-test for how to do it.

My current thinking is to pass in a lambda to RecentLookups to bump the stat, to avoid a circular dependence, since Counter depends on SymbolTable, and SymbolTable depends on RecentLookups, and RecentLookups should thus not be dependent on Counter. Might be able to finesse this with interface/impl breaks but it seems easier to use a lambda. WDYT?